pkg/controller/keystoneendpoint/keystoneendpoint_controller.go

```diff
@@ -0,0 +1,226 @@
+type ReconcileKeystoneEndpoint struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a KeystoneEndpoint object and makes changes based on the state read
+// and what is in the KeystoneEndpoint.Spec
+// TODO(user): Modify this Reconcile function to implement your Controller logic.  This example creates
+// a Pod as an example
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileKeystoneEndpoint) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling KeystoneEndpoint")
+
+	// Fetch the KeystoneEndpoint instance
+	instance := &keystonev1.KeystoneEndpoint{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if k8s_errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	reqLogger.Info("authurl", "authurl", instance.Spec.AuthURL)
+	reqLogger.Info("username", "username", instance.Spec.Username)
+	reqLogger.Info("password", "password", instance.Spec.Password)
+	reqLogger.Info("project", "project", instance.Spec.Project)
+	opts := gophercloud.AuthOptions{
+		IdentityEndpoint: instance.Spec.AuthURL,
+		Username:         instance.Spec.Username,
+		Password:         instance.Spec.Password,
+		TenantName:       instance.Spec.Project,
+		DomainName:       instance.Spec.DomainName,
+	}
+
+	provider, err := openstack.AuthenticatedClient(opts)
+	if err != nil {
+		reqLogger.Error(err, "error")
+		return reconcile.Result{}, err
+	}
+	endpointOpts := gophercloud.EndpointOpts{Type: "identity", Region: instance.Spec.Region}
+	reqLogger.Info("provider type", "%t", provider)
+	reqLogger.Info("IdentityEndpoint", "provider.IdentityEndpoint", provider.IdentityEndpoint)
+	reqLogger.Info("IdentityBase", "provider.IdentityBase", provider.IdentityBase)
+	identityClient, err := openstack.NewIdentityV3(provider, endpointOpts)
+
+	// Fetch the service ID
+	listOpts := services.ListOpts{
+		ServiceType: instance.Spec.ServiceType,
+	}
+	allPages, err := services.List(identityClient, listOpts).AllPages()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	allServices, err := services.ExtractServices(allPages)
+	reqLogger.Info("allServices", "allServices", fmt.Sprintf("%v", allServices))
```

Comment:
if there was an error, what would be held in allServices? Probably nothing good?

Reply:
this log info should be moved after the err check